src/psfmachine/perturbation.py

```diff
@@ -0,0 +1,576 @@
+        """Builds a function to lower the resolution of the data through downsampling"""
+        points = []
+        b = np.hstack([0, self.breaks, len(self.time) - 1])
+        for b1, b2 in zip(b[:-1], b[1:]):
+            p = np.arange(b1, b2, self.resolution)
+            if p[-1] != b2:
+                p = np.hstack([p, b2])
+            points.append(p)
+        points = np.unique(np.hstack(points))
+        self.nbins = len(points)
+
+        def func(x, quad=False):
+            """
+            Bins down an input variable to the same time resolution as `self`
+            """
+            if x.shape[0] == len(self.time):
+                return x[points]
+            else:
+                raise ValueError("Wrong size to bin")
+
+        return func
+
+    def _get_bindown_func(self):
+        """Builds a function to lower the resolution of the data through binning"""
+        b = np.hstack([0, self.breaks, len(self.time) - 1])
+        points = np.hstack(
+            [np.arange(b1, b2, self.resolution) for b1, b2 in zip(b[:-1], b[1:])]
+        )
+        points = points[~np.in1d(points, np.hstack([0, len(self.time) - 1]))]
+        points = np.unique(np.hstack([points, self.breaks]))
+        self.nbins = len(points) + 1
+
+        def func(x, quad=False):
+            """
+            Bins down an input variable to the same time resolution as `self`
+            """
+            if x.shape[0] == len(self.time):
+                if not quad:
+                    return np.asarray(
+                        [i.mean(axis=0) for i in np.array_split(x, points)]
+                    )
+                else:
+                    return (
+                        np.asarray(
+                            [
+                                np.sum(i ** 2, axis=0) / (len(i) ** 2)
+                                for i in np.array_split(x, points)
+                            ]
+                        )
+                        ** 0.5
+                    )
+            else:
+                raise ValueError("Wrong size to bin")
+
+        return func
+
+    def pca(self, y, ncomponents=5, smooth_time_scale=0):
+        """Adds the principal components of `y` to the design matrix
+
+        Will add two time scales of principal components, definied by `long_time_scale`
+        and `med_time_scale`.
+
+        Parameters
+        ----------
+        y: np.ndarray
+            Input flux array to take PCA of.
+        ncomponents: int
+            Number of principal components to use
+        long_time_scale: float
+            The time scale where variability is considered "long" term. Should be
+            in the same units as `self.time`.
+        med_time_scale: float
+            The time scale where variability is considered "medium" term. Should be
+            in the same units as `self.time`. Variability longer than `long_time_scale`,
+            or shorter than `med_time_scale`, will be removed before building components
```

Comment:
What is the scale of `smooth_time_scale`? 

```suggestion
        smooth_time_scale: float
             Amount to smooth the components, using a spline in time.
             If 0, the components will not be smoothed.
```